File: src/main/java/com/alexkbit/intro/ignite/service/TaskService.java
package com.alexkbit.intro.ignite.service;

import com.alexkbit.intro.ignite.model.Job;
import com.alexkbit.intro.ignite.model.Task;
import com.alexkbit.intro.ignite.model.TaskStatus;
import com.alexkbit.intro.ignite.repository.TaskRepository;
import groovy.util.logging.Slf4j;
import org.apache.ignite.IgniteQueue;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.List;

/**
 * Service for {@link Task}.
 */
@Slf4j
@Service
//@Transactional
public class TaskService {

    @Autowired
    private TaskRepository taskRepository;

    @Autowired
    private IgniteQueue<Job> jobQueue;

    public Task start(String expression) {
        Task task = new Task();
        task.setStatus(TaskStatus.WAIT);
        task.setExpression(expression);
        task = taskRepository.save(task.getId(), task);
        jobQueue.add(Job.build(task));
        return task;
    }

    public Task get(String taskId) {
        return taskRepository.findOne(taskId);
    }

    public boolean stop(String taskId) {
        Task task = taskRepository.findOne(taskId);
        if (task == null || !jobQueue.removeIf(job -> job.getTaskId().equalsIgnoreCase(taskId))) {
            return false;
        }
        task.setStatus(TaskStatus.STOPPED);
        taskRepository.save(task.getId(), task);
        return true;
    }
    public void remove(String taskId) {
        taskRepository.delete(taskId);
    }

    public Page<Task> load(int page, int count) {
        Pageable pageable = new PageRequest(page, count);
        List<Task> tasks = taskRepository.findAllOrderByCreatedAt(pageable);
        return new PageImpl<>(tasks, pageable, taskRepository.count());
    }
}

File: src/main/resources/static/js/service/taskService.js
'use strict';

App.factory('TaskService', ['$http', '$q', function($http, $q){

    var HOST = "/tasks";

    return {

        loadAllTasks: function(page, count) {
            return $http.get(HOST + '?page=' + page + '&count=' + count)
                .then(
                    function(response){
                        return response.data;
                    },
                    function(errResponse){
                        console.error('Error while load tasks');
                        return $q.reject(errResponse);
                    }
                );
        },

        createTask: function(expression){
            return $http.post(HOST + '/start', expression)
                .then(
                    function(response){
                        return response.data;
                    },
                    function(errResponse){
                        console.error('Error while start task');
                        return $q.reject(errResponse);
                    }
                );
        },

        stopTask: function(taskId){
            return $http.get(HOST + '/stop?taskId=' + taskId)
                .then(
                    function(response){
                        return response.data;
                    },
                    function(errResponse){
                        console.error('Error while stop task');
                        return $q.reject(errResponse);
                    }
                );
        },

        deleteTask: function(taskId){
            return $http.delete(HOST +'?taskId=' + taskId)
                .then(
                    function(response){
                        return response.data;
                    },
                    function(errResponse){
                        console.error('Error while deleting task');
                        return $q.reject(errResponse);
                    }
                );
        }

    };

}]);
